### ISSUE TITLE
Resource parameter for OIDC/OAuth authorize request has wrong format

### ISSUE THREAD

**Issue by leastprivilege** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#1574]
According to the resource indicator spec

https://www.rfc-editor.org/rfc/rfc8707.html#name-authorization-request

Multiple `resource` parameters are allowed.

`OpenIdConnectMessage` only allows for a single resource parameter - also the backing field is a dictionary. This needs some re-work to be spec compliant.

**Comment by leastprivilege** [comment]
Any update? This is violating the spec and blocks people from using resource indicator enabled IdPs...

**Comment by frodegil** [comment]
This has become a blocking issue for us. Our system consumes services from multiple external service providers, all using the same (external) idP. The idP uses resource indicators to authorize users for every resource/api during login so that they can request resource specific access tokens when they need to, later on.

We will try to change the Resouce property's backing field via reflection. If that (hack) doesn't work we need to use another component that allow multiple resources to be configured for the external idP. A fix, or (workaround) suggestions, will be highly appreciated.

**Comment by frodegil** [comment]
Is it possible/feasible to create a CustomOpenIdConnectMessage that inherits from OpenIdConnectMessage?
We would really like to override the CreateAuthenticationRequestUrl method so that we can build an url with multiple resource parameters (handle the "resource" parameter ourself by splitting the string and add one resource parameter per resource in a  loop inside the _parameters loop).
We've tried to clone, wrap and replace the RedirectContext.ProtocolMessage inside OnRedirectToIdentityProvider, but might have some issues with other properties not being set correctly. Not sure. It would be better if the "factory" class that creates the protocol message could create our custom message type, instead.

**Comment by frodegil** [comment]
This seems to work;

![image](https://user-images.githubusercontent.com/8426147/142275620-c610b02e-4d7a-4815-b74b-7184dcc5b7e9.png)

![image](https://user-images.githubusercontent.com/8426147/142275721-ade8e563-13d3-4bda-874b-3dc838023c60.png)

**Comment by frodegil** [comment]
Just curious, @leastprivilege . Any news on this one?

**Comment by leastprivilege** [comment]
Seems this is still not fixed.

**Comment by GavinOsborn** [comment]
+1

Multi-resource scenarios are the bedrock of the authn/authz solution for our cloud platform.
This limitation makes supporting dotnet client applications regrettably painful.

Props to @frodegil for the elegant work around 👏 but convincing software teams to adopt a solution like that can be rather troublesome when lack of official support implies that this is somehow wrong... which it isn't.

**Comment by leastprivilege** [comment]
As always, the priorities of this repo seem more around satisfying 1st party requests than fixing spec compliance.

**Comment by cakescience** [comment]
Any news about this blocker?

**Comment by leastprivilege** [comment]
Is this fixed in v7?

**Comment by jmprieur** [comment]
How would you like the API to be?
Something like:

`string[] resources`

?

**Comment by leastprivilege** [comment]
yes - and that turns into multiple `resource` parameters on the query string to the authorize endpoint.

**Comment by kyrrem** [comment]
This standard pattern will be used by many vendors in the Norwegian health care sector going forward. The national trust mechanism (HelseId) has adopted it for resource isolation between their different national APIs. With the amount of dotnet users in Norway i would expect a lot of people stumbling upon this hurdle, so it would be really nice if the oauth libraries became compliant here.

**Comment by jmprieur** [comment]
Thanks @leastprivilege
Out of curiosity why is resource used and not scopes?

**Comment by kyrrem** [comment]
For HelseId they use it as part of the resource indicator spec (https://datatracker.ietf.org/doc/html/rfc8707) as a way to handle APIs requiring to have only one audience (their audience) and to avoid having to do multiple auth rounds for getting tokens for different APIs (degrading the user experience).

More information on HelseIds developer portal:
https://helseid.atlassian.net/wiki/spaces/HELSEID/pages/481755152/Requesting+multiple+access+tokens+with+single+audiences

Why these requirements are there, and not just using scopes as you mention, i dont know. My guess would be that they tried to find a way of creating API isolation that used the most fitting oauth semantics available. 

EDIT: Im also unsure if if the specs allow changing scopes during a token request (like refresh)

**Comment by leastprivilege** [comment]
> Thanks @leastprivilege
> Out of curiosity why is resource used and not scopes?

Resources and scopes are orthogonal - e.g. multiple resources could offer the same scopes but you still want to be able to restrict the access to a specific resource (a common scenario for e.g. in SMART on FHIR).

Also resources might live in different trust domains - e.g. resourceA should not be able to forward incoming tokens to resourceB - with the additional resource identifier, resourceB would not accept the token.

And lastly this is a way to prevent overly powerful access tokens by forcing the client to use per-resource access tokens (that can have per-resource configuration like claims or encryption keys/methods)

See here for details
https://datatracker.ietf.org/doc/html/rfc8707

**Comment by frodegil** [comment]
The previous "hack" worked when the AuthenticationMethod was (default) set to **RedirectGet**.

When adding support for DPoP, the (authorize) state parameter and the query-string became very long, resulting in a HTTP status 404 on the authorization server (Login).

The workaround was to change the AuthenticationMethod to **FormPost**

The code I posted earlier works for RedirectGet only. You have to overide the OpenIdConnectMessage's BuildFormPost method to build a post that supports multiple resources;

![image](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/assets/8426147/83214724-5d31-452c-89da-9be7e48e87dd)

**Comment by frodegil** [comment]
For others reading this thread.

Duende IdentityModel (we're using 7.0.0.0) and the PushedAuthorizationRequest class supports multiple resources. 
We have replaced the old form-post and query-string hacks, mentioned in this thread, with PAR.
It works like a charm.
